### PR TITLE
Add message attributes to transactions and spans

### PR DESCRIPTION
**Pull request by simitt**
Add attributes holding messaging information to events to support monitoring message systems.

**Intake API:**
All message attributes defined in https://github.com/elastic/apm/issues/143#issue-491632886 are added to the Intake API: 

`message.queue.name` (string, maxLength 1024, ~required~)
~`message.topic.name` (string, maxLength 1024, required)~
`message.age.ms` (integer, optional)
`message.body` (string, optional)
`message.headers` (object with string or array properties, optional)

There is no distinction between `transaction` and `span` messages as there is no obvious advantage of separate handling, while a distinction would duplicate certain fields and therefore be more error prone. 

~If `span.type == messaging || transaction.type == messaging` and no message information is sent an error will be thrown. The check for this is not in the json spec but in the go code.~  

**Storage ES**
~Since ECS already defines a root level `message` field as `text`, a root level object `messaging` is introduced. To allow aggregations and easy queries over message fields for spans and transactions at the same time, the information is not stored under the event type `transaction` and `span` but under a root level key `messaging`.  This object `messaging` contains a `type` defining the messaging system, e.g. `JMS, rabbitmq`, etc. and the actual `message`.~

~For `spans` the `span.subtype` is copied to `messaging.type`, for `transactions` this field is  unknown.~
~For `spans` the `span.action` is copied to `messaging.message.operation`, for `transactions` this field is unknown.~
~The fields are copied for easier queries when this information is needed.~

Since there is no immediate use case for aggregating messaging related information, the information is stored under the event type, to avoid introducing a non-ECS root field. 
Information is stored as `transaction.message` and `span.message`. 

Following fields are indexed:
* ~`messaging.type`~
* ~`messaging.message.queue.name`~
* ~`messaging.message.topic.name`~
* ~`messaging.message.age.ms`~
* ~`messaging.message.operation`~
* `span.message.queue.name`
* `span.message.age.ms`
* `transaction.message.queue.name`
* `transaction.message.age.ms`

**Message Headers**
The message headers are internally treated the same way as http headers, which means that they are canonicalized before stored to Elasticsearch. 

implements #2697 

@eyalkoren please let me know if this makes sense to you.

### CODE REVIEW

**Comment by eyalkoren** [comment]
LGTM, but I am not fully aware of the implications of storage decisions.
The best input I can provide is that it can be useful to look at the relationships between messaging spans and transactions as (mostly) equivalent to the relationships between inbound (Transaction) and outbound (Span) HTTP requests. 
For example:
> If `span.type == messaging || transaction.type == messaging` and no message information is sent an error will be thrown. The check for this is not in the json spec but in the go code.

Do you do the same with HTTP spans/transactions sent without request info?
Do you also store an equivalent to `messaging.type` for HTTP (eg the name of an HTTP client related to an HTTP span)? What would it be used for?

Another example is that only `messaging` transactions can contain header or body info, similar to HTTP transactions vs. spans, if that makes any difference.

Lastly- you should be aware that `message.queue.name` and `message.topic.name` should not be sent for the same message. It is either one or the other. The reason we decided to have both is that either would make sense for different systems/scenarios.

**Comment by simitt** [comment]
>> If span.type == messaging || transaction.type == messaging and no message information is sent an error will be thrown. The check for this is not in the json spec but in the go code.

> Do you do the same with HTTP spans/transactions sent without request info?

The check was added based on the comment in https://github.com/elastic/apm/issues/143#issue-491632886: _context.message.queue.name and context.message.topic.name: required for messaging spans and transactions. Will be used for the service map. Indexed as keyword._

> Do you do the same with HTTP spans/transactions sent without request info?

No we don't. But according to the description I referred to above, a message information is required when the event type is `messaging`. I can loosen the requirement, but the UI and service map logic need to be able to handle missing `messaging` information then.
@eyalkoren I am fine either way; @elastic/apm-ui can you let us know if having an `event.type=messaging` without any `messaging` information would be problematic?

According to @eyalkoren 's comment above, I will change the json spec requirement to only require either `message.queue.name` or `message.topic.name` when a `message` is sent. 

> Do you also store an equivalent to messaging.type for HTTP (eg the name of an HTTP client related to an HTTP span)? What would it be used for?

My reasoning was that one might want to filter by `messaging.type`, e.g. when using `JMS` but also something else like `rabbitmq`. This was mainly thought for querying convenience for the UI. But if that field doesn't seem important I can remove it. @elastic/apm-ui maybe we should do a quick sync on how the data will be used.

**Comment by eyalkoren** [comment]
> The check was added based on the comment in elastic/apm#143 (comment): context.message.queue.name and context.message.topic.name: required for messaging spans and transactions. Will be used for the service map. Indexed as keyword.

Sorry, that was long ago, updated it. New dedicated `destination.service` fields were added for the service map. Those are nested within the top `context` level, so you can enforce their existence for the purpose of service maps as you would do with others.

I am really fine with any decision based on storage/query convenience, my comments are only for making sure you are aware of what they mean from the agent perspective.

**Comment by simitt** [comment]
From the updated comment:
> context.message.queue.name and context.message.topic.name: required for messaging spans and transactions. Indexed as keyword.

@eyalkoren I changed the Intake API to either require `context.message.queue.name` OR `context.message.topic.name` to be present if `context.message.*` information is sent. Is this how it should be or should none of this information be required?

**Comment by eyalkoren** [comment]
@simitt right, didn't update that 🤦‍♂ 
Updated now- let's make it optional. The reason is that there are cases where we trace queue/topic polling actions that happen within a traced transaction, and those may be reported without a queue/topic.

**Comment by simitt** [comment]
@eyalkoren are the values for `span.action` fixed to `receive` and `send`? I currently copy the `span.action` to the `messaging.message.operation` field, and think it would make sense to also set it for `transaction` events. 

By storing the messaging related information under `messaging` instead of `event_type.message` and duplicating the action, queries can be simplified. 
E.g. to query for all incoming messages for one trace the query would be `trace_id=xyz AND messaging.message.operation=receive` rather than `trace_id=xyz AND (span.type=messaging AND span.action=receive OR transaction.type=messaging)`.

**Comment by eyalkoren** [comment]
> are the values for span.action fixed to receive and send?

No, I use `send` and `poll` for Kafka... It may still make sense to use `receive` for this purpose, as `poll` wouldn't have the `age` filed, only `receive` spans and transactions.

Honestly, I wouldn't drive any schema decisions if the only query it is required for is the `age` field. This is a somewhat experimental/evaluation info that was recently added and I prefer we can get some milage with it and evaluate it.

**Comment by simitt** [comment]
After discussions with @eyalkoren and @sqren I made following changes:
* store `message` information under event type, e.g. `transaction.message` and `span.message`. There is no use case planned for aggregating on the message information and it is also not part of service maps. The only currently planned query is for visualizing message information in the `trace metadata`. Therefore introducing a dedicated (non ECS) root field for potential future query or aggregation use cases seems premature.
* do not require any `message` fields on the Intake API
* remove check that `message` information is sent when `span.type=messaging` or `transaction.type=messaging`, as there might be valid use cases for this.
* remove `context.message.topic.name`, the information will be sent via `context.message.queue.name`. 
* fix system tests

There are some uncertainties about how the message information should be used from an UI perspective and mid- to longterm plans with that. Maybe @eyalkoren or @nehaduggal can bring more clarification into this from a product view, to avoid driving the discussion from an implementation viewpoint (might be better to clarify and move this discussion to https://github.com/elastic/kibana/issues/49465 or https://github.com/elastic/apm/issues/143).

**Comment by simitt** [comment]
According to https://github.com/elastic/kibana/issues/49465#issuecomment-571183457 it seems the UI team is fine with the suggested ES storage.

File: model/message.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/elastic/beats/libbeat/common"
+
+	"github.com/elastic/apm-server/utility"
+)
+
+// Message holds information about a recorded message, such as the message body and meta information
+type Message struct {
+	Body        *string
+	Headers     http.Header
+	AgeMicroSec *int
```
[the remark this answers is not in the document]
Comment (simitt) [reply model/message.go]: right 🤦‍♀

File: beater/test_approved_es_documents/TestPublishIntegrationSpans.approved.json [hunk 2]
```diff
@@ -373,6 +373,67 @@
...
+                "name": "backendspans"
+            },
+            "span": {
+                "action": "receive",
+                "duration": {
+                    "us": 141581
+                },
+                "id": "00xxx12312312312",
+                "message": {
+                    "age.ms": 1577958057123,
```
[the remark this answers is not in the document]
Comment (simitt) [reply beater/test_approved_es_documents/TestPublishIntegrationSpans.approved.json]: There is no functional difference, but I aligned it now.

File: utility/map_str_enhancer.go [hunk 3]
```diff
@@ -131,6 +132,12 @@ func update(m common.MapStr, key string, val interface{}, remove bool) {
 		m[key] = val
 	case int, int8, int16, int32, int64, uint, uint8, uint32, uint64:
 		m[key] = val
+	case http.Header:
+		if value != nil {
+			m[key] = value
+		} else if remove {
+			delete(m, key)
+		}
```
[the remark this answers is not in the document]
Comment (simitt) [reply utility/map_str_enhancer.go]: > all the else if branches that only check for (val == nil && remove) (6 occurrences) are already covered by L:47-52:

The checks are necessary as a non-nil interface can point to a nil value, see https://play.golang.com/p/nhsiTGDhg7V

[the remark this answers is not in the document]
Comment (mdelapenya) [reply utility/map_str_enhancer.go]: Ahhh sorry, I committed a mistake when reading the code: the internal evaluation is for `value`, not for `val`, so it's checking the type (`switch value := val.(type)`)

Thanks for the clarification!

**Pull request merged**